include/unifex/async_scope.hpp

```diff
@@ -561,6 +564,201 @@ struct future final {
 
+template <typename Operation, typename Receiver>
+struct _cleaning_receiver final {
+  struct type;
+};
+
+template <typename Operation, typename Receiver>
+using cleaning_receiver =
+    typename _cleaning_receiver<Operation, Receiver>::type;
+
+template <typename Operation, typename Receiver>
+struct _cleaning_receiver<Operation, Receiver>::type final {
+  template <typename... Values>
+  void set_value(Values&&... values) noexcept {
+    op_->cleanup();
+    unifex::set_value(std::move(receiver_), (Values &&) values...);
+  }
+
+  template <typename E>
+  void set_error(E&& e) noexcept {
+    op_->cleanup();
+    unifex::set_error(std::move(receiver_), (E &&) e);
+  }
+
+  void set_done() noexcept {
+    op_->cleanup();
+    unifex::set_done(std::move(receiver_));
+  }
+
+  template(typename CPO)                       //
+      (requires is_receiver_query_cpo_v<CPO>)  //
+      friend auto tag_invoke(CPO cpo, const type& r) noexcept(
+          noexcept(static_cast<CPO&&>(cpo)(r.receiver_)))
+          -> callable_result_t<CPO, const Receiver&> {
+    return static_cast<CPO&&>(cpo)(r.receiver_);
+  }
+
+  friend inplace_stop_token
+  tag_invoke(tag_t<get_stop_token>, const type& r) noexcept {
+    return r.op_->get_token();
+  }
+  Operation* op_;
+  UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+};
+
+template <typename Sender, typename Receiver>
+struct _attached_op final {
+  class type;
+};
+
+template <typename Sender, typename Receiver>
+using attached_operation =
+    typename _attached_op<Sender, remove_cvref_t<Receiver>>::type;
+
+template <typename Sender, typename Receiver>
+class _attached_op<Sender, Receiver>::type final {
+  using cleaning_receiver_t = cleaning_receiver<type, Receiver>;
+  using nested_operation_t = connect_result_t<Sender, cleaning_receiver_t>;
+
+public:
+  template <typename Receiver2>
+  explicit type(Sender&& s, Receiver2&& r, async_scope* scope) noexcept(
+      is_nothrow_connectable_v<Sender, cleaning_receiver_t>&& std::
+          is_nothrow_constructible_v<cleaning_receiver_t, type*, Receiver2>&&
+              std::is_nothrow_constructible_v<Receiver, Receiver2>)
+    : scope_((std::uintptr_t)scope & mask)
+    , receiverToken_(get_stop_token(r)) {
+    if (scope_ref()) {
+      unifex::activate_union_member_with(op_, [&] {
+        return unifex::connect(
+            (Sender &&) s, cleaning_receiver_t{this, (Receiver2 &&) r});
+      });
+    } else {
+      unifex::activate_union_member(receiver_, (Receiver2 &&) r);
+    }
+  }
+
+  type(type&& op) = delete;
+
+  ~type() {
+    if (scope_ref()) {
+      unifex::deactivate_union_member(op_);
+      // started => receiver responsible for cleanup
+      if (!has_started()) {
+        cleanup();
+      }
+    } else {
+      unifex::deactivate_union_member(receiver_);
+    }
+  }
+
+  auto get_token() { return stopSource_.get_token(); }
+
+  void cleanup() noexcept {
+    UNIFEX_ASSERT(scope_ref() != nullptr);
+    stopSource_.deregister_callbacks();
```

Comment:
Should we do this unconditionally?  I _think_ it's OK because it's invoking `optional<>::reset()` under the covers, which is safe if the optional's already empty.

Reply:
That was my reasoning as well. Rather than introducing another branch, that obfuscates control flow.